include/openthread/history_tracker.h

```diff
@@ -0,0 +1,211 @@
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *
+ * @brief
+ *   This file includes the OpenThread API for History Tracker module.
+ */
+
+#ifndef OPENTHREAD_HISTORY_TRACKER_H_
+#define OPENTHREAD_HISTORY_TRACKER_H_
+
+#include <openthread/instance.h>
+#include <openthread/ip6.h>
+#include <openthread/thread.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief
+ *   This header defines the public API for History Tracker.
+ *
+ *   History Tracker module records history of different events (e.g. RX and TX messages or network info changes, etc.)
+ *   as the Thread network operates. All tracked entries are timestamped.
+ *
+ *   The functions in this module are available when `OPENTHREAD_CONFIG_HISTOR_TRACKER_ENABLE` is enabled.
+ *
+ */
+
+/**
+ * This constant specifies the maximum age of entries which is 49 days (in msec).
+ *
+ * Entries older than the max age will give this value as their age.
+ *
+ */
+#define OT_HISTORY_TRACKER_MAX_AGE (49 * 24 * 60 * 60 * 1000u)
+
+/**
+ * This type represents an iterator to iterate through a history list.
+ *
+ * The fields in this type are opaque (intended for use by OpenThread core) and therefore should not be accessed/used
+ * by caller.
+ *
+ * Before using an iterator, it MUST be initialized using `otHistoryTrackerInitIterator()`,
+ *
+ */
+typedef struct otHistoryTrackerIterator
+{
+    uint32_t mData32;
+    uint16_t mData16;
+} otHistoryTrackerIterator;
+
+/**
+ * This structure represents Thread network info.
+ *
+ */
+typedef struct otHistoryTrackerNetworkInfo
+{
+    otDeviceRole     mRole;        ///< Device Role.
+    otLinkModeConfig mMode;        ///< Device Mode.
+    uint16_t         mRloc16;      ///< Device RLOC16.
+    uint32_t         mPartitionId; ///< Partition ID (valid when attached).
+} otHistoryTrackerNetworkInfo;
+
+/**
+ * Constants representing message priority used in `otHistoryTrackerMessageInfo` struct.
+ *
+ */
+enum
+{
+    OT_HISTORY_TRACKER_MSG_PRIORITY_LOW    = OT_MESSAGE_PRIORITY_LOW,      ///< Low priority level.
+    OT_HISTORY_TRACKER_MSG_PRIORITY_NORMAL = OT_MESSAGE_PRIORITY_NORMAL,   ///< Normal priority level.
+    OT_HISTORY_TRACKER_MSG_PRIORITY_HIGH   = OT_MESSAGE_PRIORITY_HIGH,     ///< High priority level.
+    OT_HISTORY_TRACKER_MSG_PRIORITY_NET    = OT_MESSAGE_PRIORITY_HIGH + 1, ///< Network Control priority level.
+};
+
+/**
+ * This structure represents a RX/TX IPv6 message info.
+ *
+ * Some of the fields in this struct are applicable to a RX message or a TX message only, e.g., `mAveRxRss` is the
+ * average RSS of all fragment frames that form a received message and is only applicable for a RX message.
+ *
+ */
+typedef struct otHistoryTrackerMessageInfo
```

Comment:
We usually only used `PACKED` for types that may be added in in message. 
The fields in this `struct` are already defined in an order to try to avoid alignment gaps.

In general, the `HistoryTracker` is not really intended for use on devices that have resource limitation. So memory use should not really considered critical here.